src/ext/db/connection.cr

```diff
@@ -1,8 +1,66 @@
 module DB
   abstract class Connection
+    @expires_at : Time?
+    @ttl : Int32 = 0
+    @retry_delay : Int32 = 1
+    @reaper : Tasker::OneShot(Nil)?
+
+    property ttl, expires_at, reaper, retry_delay
+
     # :nodoc:
     getter _avram_stack = [] of DB::Transaction
 
+    # :nodoc:
+    def expired?
+      @expires_at.try &.<= Time.utc
+    end
+
+    def schedule_reaper!(pool : DB::Pool)
+      return if @reaper || ttl.zero?
+      ::Log.info { "scheduling reaper for #{object_id}, #{expires_at}" }
+
+      self.reaper ||= Tasker.at(expires_at.not_nil!) { reap!(pool, ttl) }
+    end
+
+    def reap!(pool : DB::Pool, ttl : Int32) : Nil
+      reaped = false
+
+      ::Log.info { "reaping #{object_id}" }
+
+      pool.each_resource do |cnn|
+        if cnn.object_id == object_id
+          ::Log.info { "closing #{cnn.object_id}" }
+          reaped = true
+          cnn.close
+          spawn { pool.create_expiring_connection!(ttl, retry_delay) }
+        end
+      end
```

Comment:
`Pool#close_resource(cnn)`

Reply:
Will check the source, but that makes sure your connection is in the @idle pool?